src/ModelContextProtocol/Protocol/Transport/InMemoryTransport.cs

```diff
@@ -0,0 +1,416 @@
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Logging.Abstractions;
+
+using ModelContextProtocol.Configuration;
+using ModelContextProtocol.Logging;
+using ModelContextProtocol.Protocol.Messages;
+using ModelContextProtocol.Server;
+
+using System.Diagnostics.CodeAnalysis;
+using System.Threading.Channels;
+
+
+namespace ModelContextProtocol.Protocol.Transport;
+
+/// <summary>
+/// Provides an in memory implementation of the MCP transport protocol over shared memory.
+/// This transport enables efficient in-process MCP functionality and easier test authoring.
+/// </summary>
+/// <remarks>
+/// <para>
+/// The InMemoryTransport allows both client and server to communicate within the same process,
+/// which is particularly useful for testing scenarios and embedded applications.
+/// </para>
+/// <para>
+/// This implementation requires dynamic code access for tool registration and might not work in Native AOT.
+/// </para>
+/// </remarks>
+[RequiresUnreferencedCode(RequiresUnreferencedCodeMessage)]
+public sealed class InMemoryTransport : TransportBase, IServerTransport, IClientTransport
+{
+    private const string RequiresUnreferencedCodeMessage = "This method requires dynamic lookup of method metadata and might not work in Native AOT.";
```

Comment:
What is this using that requires this?

Reply:
that was for a warning coming from the call to `WithTools`

Reply:
Why does the transport need tools? Looking at the implementation further, I'm actually confused now about what it's doing... why is it creating a server? Wouldn't the transport be passed to McpServerFactory.Create rather than the transport creating servers?

@halter73, does this get back to the notion of a server erroneously being both 1:1 and 1:many?

Reply:
That code is all gone

Reply:
> does this get back to the notion of a server erroneously being both 1:1 and 1:many?

Yes. This test transport really only works 1:1 because it implements `IServerTransport`. However, you could essentially make multiple instances of this to simulate multiple connections.